File: cybertron/shared/src/protocol.ts
/**
 * Shared wire protocol between the runtime gateway (runtime/src/server.ts),
 * the Electron preload bridge (electron/preload.ts), and the UI (app/, cli).
 *
 * This is the one file every side imports from, so the contract can't drift.
 */

export type ToolCategory = "recon" | "crawl" | "scan" | "secrets" | "defense" | "exploit";
// "defense" = blue-team tooling (detection/response). "exploit" = red-team
// verification steps, hard-gated behind human approval regardless of team.

export interface ToolDefinition {
  id: string;
  category: ToolCategory;
  label: string;
  /** true = runs immediately when the agent calls it; false = needs explicit human approval first */
  autoApprove: boolean;
  implemented: boolean;
}

export interface ToolCallRequest {
  type: "tool_call_request";
  sessionId: string;
  requestId: string;
  toolId: string;
  args: Record<string, string>;
}

export interface ToolCallApproval {
  type: "tool_call_approval";
  sessionId: string;
  requestId: string;
  approved: boolean;
}

export interface ToolCallResult {
  type: "tool_call_result";
  sessionId: string;
  requestId: string;
  toolId: string;
  ok: boolean;
  output: string;
  error?: string;
  durationMs: number;
}

export type AgentState = "idle" | "thinking" | "awaiting_approval" | "running_tool" | "done" | "error";

export type SessionOrigin = "web" | "cli";

export interface AgentStatusEvent {
  type: "agent_status";
  sessionId: string;
  state: AgentState;
  detail?: string;
  /** cumulative tokens used by this session so far, if the model returned usage data */
  tokensUsed?: number;
  /** the model's context window, passed through so the client never has to hardcode it */
  contextWindow?: number;
}

export interface AgentSessionStart {
  type: "session_start";
  /** client-generated id, so a session can be tracked before the server acks it */
  sessionId: string;
  goal: string;
  origin: SessionOrigin;
}

export interface AuthCommand {
  type: "auth";
  token: string;
}

export interface AuthResultEvent {
  type: "auth_result";
  ok: boolean;
}

export interface ListSessionsCommand {
  type: "list_sessions";
}

export interface GetConfigCommand {
  type: "get_config";
}

export interface SetConfigCommand {
  type: "set_config";
  nimApiKey: string;
}

export interface ConfigStateEvent {
  type: "config_state";
  /** never the actual key value — only whether one is set, and where it
   * came from, so the settings panel can show real status without ever
   * receiving the secret back down over the wire once it's saved. */
  nimApiKeySet: boolean;
  nimApiKeySource: "env" | "file" | "none";
}

export interface GetToolsCommand {
  type: "get_tools";
}

export interface ToolsCatalogEvent {
  type: "tools_catalog";
  tools: ToolDefinition[];
}

export interface SessionSummary {
  id: string;
  goal: string;
  state: AgentState;
  startedAt: number;
  finishedAt?: number;
  toolCallCount: number;
  lastToolId?: string;
  origin: SessionOrigin;
}

export interface SessionsSnapshotEvent {
  type: "sessions_snapshot";
  sessions: SessionSummary[];
}

export type RuntimeEvent =
  | ToolCallRequest
  | ToolCallResult
  | AgentStatusEvent
  | SessionsSnapshotEvent
  | AuthResultEvent
  | ConfigStateEvent
  | ToolsCatalogEvent;

export type RuntimeCommand =
  | AgentSessionStart
  | ToolCallApproval
  | ListSessionsCommand
  | AuthCommand
  | GetConfigCommand
  | SetConfigCommand
  | GetToolsCommand;

export const GATEWAY_PORT = 8765;
// Explicit IP, not "localhost" — the hostname form is what a system-wide
// proxy (common on Kali if Burp Suite or similar is configured) or a
// misbehaving resolver can intercept/redirect, even for loopback traffic.
// Plain Node (the TUI, the `ws` client) doesn't consult a browser's proxy
// settings, but Electron's Chromium-based renderer does — which is why the
// TUI could connect fine while the desktop UI hung on "Connecting...".
export const GATEWAY_HOST = "127.0.0.1";

File: cybertron (1)/cybertron/src/core/protocol.py
"""
Cybertron Python UI — Protocol definitions matching the Node.js runtime exactly.
Extended with GitHub Tool Loader message types.
"""
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Any
from enum import Enum


class AgentState(str, Enum):
    IDLE = "idle"
    THINKING = "thinking"
    RUNNING_TOOL = "running_tool"
    AWAITING_APPROVAL = "awaiting_approval"
    DONE = "done"
    ERROR = "error"


class MessageType(str, Enum):
    # Auth & Session
    AUTH = "auth"
    AUTH_RESULT = "auth_result"
    SESSION_START = "session_start"
    SESSION_STARTED = "session_started"
    # Agent lifecycle
    AGENT_STATUS = "agent_status"
    # Tool execution
    TOOL_CALL_REQUEST = "tool_call_request"
    TOOL_CALL_APPROVAL = "tool_call_approval"
    TOOL_CALL_RESULT = "tool_call_result"
    # Server state
    SESSIONS_SNAPSHOT = "sessions_snapshot"
    LIST_SESSIONS = "list_sessions"
    # Tool catalog
    GET_TOOLS = "get_tools"
    TOOLS_CATALOG = "tools_catalog"
    # Config
    CONFIG_STATE = "config_state"
    # ── GitHub Tool Loader ──────────────────────────────────────────────────
    REGISTER_GITHUB_TOOL = "register_github_tool"
    GITHUB_TOOL_STATUS = "github_tool_status"
    REMOVE_TOOL = "remove_tool"
    GET_MARKETPLACE = "get_marketplace"
    MARKETPLACE_CATALOG = "marketplace_catalog"
    # ── Streaming / Intelligence ───────────────────────────────────────────
    STREAM_TOKEN = "stream_token"
    AGENT_PLAN = "agent_plan"
    # ── Control Center ────────────────────────────────────────────────────
    GET_CONFIG = "get_config"
    SET_CONFIG = "set_config"
    CONFIG_UPDATED = "config_updated"
    # ── Dry Run ────────────────────────────────────────────────────────────
    DRY_RUN = "dry_run"
    DRY_RUN_RESULT = "dry_run_result"


GATEWAY_HOST = "127.0.0.1"
GATEWAY_PORT = 8765


@dataclass
class SessionSummary:
    id: str
    goal: str
    state: str
    startedAt: int
    finishedAt: Optional[int] = None
    toolCallCount: int = 0
    lastToolId: Optional[str] = None
    origin: Optional[str] = None


@dataclass
class ToolCatalogEntry:
    id: str
    label: str
    category: str
    autoApprove: bool
    implemented: bool


@dataclass
class PendingApproval:
    sessionId: str
    requestId: str
    toolId: str
    args: Dict[str, Any]


@dataclass
class GitHubToolSpec:
    id: str
    repo: str
    label: str
    category: str
    binary_name: str
    binary_path: str
    version: str
    auto_approve: bool = False
    handler: str = "github"
    installed_at: str = ""
    last_updated: str = ""
    schema: Dict[str, Any] = field(default_factory=dict)
